### PR TITLE
Proxy support select statement '*'

### LINKED ISSUE #11202: Connect proxy failure if using driver of MySQL 8 and prepared statement
When I used driver `8.0.11` of MySQL and enabled prepared statement(using `&useServerPrepStmts=true&cachePrepStmts=true` in JDBC URL) to connect proxy, the exception occur:

```log
java.sql.SQLException: java.lang.ArrayIndexOutOfBoundsException: 1
	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:127)
	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:95)
	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:87)
	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:61)
	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:71)
	at com.mysql.cj.jdbc.ServerPreparedStatement.executeInternal(ServerPreparedStatement.java:420)
	at com.mysql.cj.jdbc.ClientPreparedStatement.executeQuery(ClientPreparedStatement.java:1019)
	at Main.executeQuery(Main.java:51)
	at Main.main(Main.java:41)
Caused by: java.lang.ArrayIndexOutOfBoundsException: 1
	at com.mysql.cj.protocol.a.MergingColumnDefinitionFactory.createFromFields(MergingColumnDefinitionFactory.java:61)
	at com.mysql.cj.protocol.a.ColumnDefinitionReader.read(ColumnDefinitionReader.java:80)
	at com.mysql.cj.protocol.a.ColumnDefinitionReader.read(ColumnDefinitionReader.java:40)
	at com.mysql.cj.protocol.a.NativeProtocol.read(NativeProtocol.java:1648)
	at com.mysql.cj.protocol.a.BinaryResultsetReader.read(BinaryResultsetReader.java:70)
	at com.mysql.cj.protocol.a.BinaryResultsetReader.read(BinaryResultsetReader.java:50)
	at com.mysql.cj.protocol.a.NativeProtocol.read(NativeProtocol.java:1661)
	at com.mysql.cj.protocol.a.NativeProtocol.readAllResults(NativeProtocol.java:1707)
	at com.mysql.cj.ServerPreparedQuery.readExecuteResult(ServerPreparedQuery.java:419)
	at com.mysql.cj.ServerPreparedQuery.serverExecute(ServerPreparedQuery.java:202)
	at com.mysql.cj.jdbc.ServerPreparedStatement.serverExecute(ServerPreparedStatement.java:620)
	at com.mysql.cj.jdbc.ServerPreparedStatement.executeInternal(ServerPreparedStatement.java:397)
	... 3 more
```

But driver `5.1.42` of MySQL is fine.

The test class is:

```java
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.text.DecimalFormat;

public final class Main {
    
    public static void main(final String[] args) throws ClassNotFoundException {
        String driverClassName = "com.mysql.jdbc.Driver";
        String url = "jdbc:mysql://localhost:3307/readwrite_splitting_db?serverTimezone=UTC&useSSL=false&useUnicode=true&characterEncoding=UTF-8&useServerPrepStmts=true&cachePrepStmts=true";
        String username = "root";
        String password = "root";
        String sql = "SELECT * FROM t_order WHERE order_id=?";
        Class.forName(driverClassName);
        try (Connection connection = DriverManager.getConnection(url, username, password);
             PreparedStatement preparedStatement = connection.prepareStatement(sql)
        ) {
            for (int i = 0; i < 10; i++) {
                executeQuery(preparedStatement);
            }
        } catch (final SQLException ex) {
            ex.printStackTrace();
        }
    }
    
    private static void executeQuery(final PreparedStatement preparedStatement) throws SQLException {
        preparedStatement.setLong(1, 1);
        try (ResultSet resultSet = preparedStatement.executeQuery()) {
            while (resultSet.next()) {
                resultSet.getObject(1);
            }
        }
    }
}

```

Configuration of config-readwrite-splitting.yaml

```yaml
schemaName: readwrite_splitting_db

dataSources:
  write_ds:
    url: jdbc:mysql://127.0.0.1:3306/demo_write_ds?serverTimezone=UTC&useSSL=false
    username: root
    password:
    connectionTimeoutMilliseconds: 30000
    idleTimeoutMilliseconds: 60000
    maxLifetimeMilliseconds: 1800000
    maxPoolSize: 50
    minPoolSize: 1
    maintenanceIntervalMilliseconds: 30000
  read_ds_0:
    url: jdbc:mysql://127.0.0.1:3306/demo_read_ds_0?serverTimezone=UTC&useSSL=false
    username: root
    password:
    connectionTimeoutMilliseconds: 30000
    idleTimeoutMilliseconds: 60000
    maxLifetimeMilliseconds: 1800000
    maxPoolSize: 50
    minPoolSize: 1
    maintenanceIntervalMilliseconds: 30000
  read_ds_1:
    url: jdbc:mysql://127.0.0.1:3306/demo_read_ds_1?serverTimezone=UTC&useSSL=false
    username: root
    password:
    connectionTimeoutMilliseconds: 30000
    idleTimeoutMilliseconds: 60000
    maxLifetimeMilliseconds: 1800000
    maxPoolSize: 50
    minPoolSize: 1
    maintenanceIntervalMilliseconds: 30000

rules:
- !READWRITE_SPLITTING
  dataSources:
    pr_ds:
      writeDataSourceName: write_ds
      readDataSourceNames:
        - read_ds_0
        - read_ds_1
```

Configuration of server.yaml

```yaml

rules:
  - !AUTHORITY
    users:
      - root@%:root
      - sharding@:sharding
    provider:
      type: NATIVE

props:
  max-connections-size-per-query: 1
  executor-size: 16  # Infinite by default.
  proxy-frontend-flush-threshold: 128  # The default value is 128.
    # LOCAL: Proxy will run with LOCAL transaction.
    # XA: Proxy will run with XA transaction.
    # BASE: Proxy will run with B.A.S.E transaction.
  proxy-transaction-type: LOCAL
  xa-transaction-manager-type: Atomikos
  proxy-opentracing-enabled: false
  proxy-hint-enabled: false
  sql-show: false
  check-table-metadata-enabled: false
  lock-wait-timeout-milliseconds: 50000 # The maximum time to wait for a lock
  show-process-list-enabled: false
```

### PR DESCRIPTION
**Pull request by donjuan6**
Fixes #11202 .

Changes proposed in this pull request:
- Proxy support select statement '*'

### CODE REVIEW

File: shardingsphere-proxy/shardingsphere-proxy-frontend/shardingsphere-proxy-frontend-mysql/src/main/java/org/apache/shardingsphere/proxy/frontend/mysql/command/query/binary/prepare/MySQLComStmtPrepareExecutor.java [hunk 1]
```diff
@@ -66,9 +73,41 @@
     }
     
     private int getProjectionCount(final SQLStatement sqlStatement) {
-        return sqlStatement instanceof SelectStatement ? ((SelectStatement) sqlStatement).getProjections().getProjections().size() : 0;
+        return sqlStatement instanceof SelectStatement
+                ? getProjectionCount((SelectStatement) sqlStatement) : 0;
```
Comment (tristaZero) [inline shardingsphere-proxy/shardingsphere-proxy-frontend/shardingsphere-proxy-frontend-mysql/src/main/java/org/apache/shardingsphere/proxy/frontend/mysql/command/query/binary/prepare/MySQLComStmtPrepareExecutor.java:77]: Hi could you move this relevant handling to `ProjectionsContext` which is a suitable place for this info.

File: shardingsphere-proxy/shardingsphere-proxy-frontend/shardingsphere-proxy-frontend-mysql/src/main/java/org/apache/shardingsphere/proxy/frontend/mysql/command/query/binary/prepare/MySQLComStmtPrepareExecutor.java [hunk 2]
```diff
@@ -66,9 +73,41 @@
...
+    }
+
+    private boolean isSimpleTableStatement(SelectStatement statement) {
+        return statement.getFrom() instanceof SimpleTableSegment;
+    }
+
+    private int getTableColumnCount(SimpleTableSegment table) {
+        String schemaName = backendConnection.getSchemaName();
+        ShardingSphereMetaData shardingSphereMetaData = ProxyContext.getInstance().getMetaData(schemaName);
+
```
Comment (tristaZero) [inline shardingsphere-proxy/shardingsphere-proxy-frontend/shardingsphere-proxy-frontend-mysql/src/main/java/org/apache/shardingsphere/proxy/frontend/mysql/command/query/binary/prepare/MySQLComStmtPrepareExecutor.java:103]: Please remove these redundant blank lines.

**Review by tristaZero — CHANGES_REQUESTED** [review]
Thanks for this PR, please look at my comments.

File: shardingsphere-proxy/shardingsphere-proxy-frontend/shardingsphere-proxy-frontend-mysql/src/main/java/org/apache/shardingsphere/proxy/frontend/mysql/command/query/binary/prepare/MySQLComStmtPrepareExecutor.java (hunk 1 above)
Comment (donjuan6) [reply shardingsphere-proxy/shardingsphere-proxy-frontend/shardingsphere-proxy-frontend-mysql/src/main/java/org/apache/shardingsphere/proxy/frontend/mysql/command/query/binary/prepare/MySQLComStmtPrepareExecutor.java:77]: I'm sorry, I read the Proxy code for the first time, and I don’t understand how to move it.

File: shardingsphere-proxy/shardingsphere-proxy-frontend/shardingsphere-proxy-frontend-mysql/src/main/java/org/apache/shardingsphere/proxy/frontend/mysql/command/query/binary/prepare/MySQLComStmtPrepareExecutor.java (hunk 2 above)
Comment (donjuan6) [reply shardingsphere-proxy/shardingsphere-proxy-frontend/shardingsphere-proxy-frontend-mysql/src/main/java/org/apache/shardingsphere/proxy/frontend/mysql/command/query/binary/prepare/MySQLComStmtPrepareExecutor.java]: ok, done!

File: shardingsphere-proxy/shardingsphere-proxy-frontend/shardingsphere-proxy-frontend-mysql/src/main/java/org/apache/shardingsphere/proxy/frontend/mysql/command/query/binary/prepare/MySQLComStmtPrepareExecutor.java (hunk 1 above)
Comment (donjuan6) [reply shardingsphere-proxy/shardingsphere-proxy-frontend/shardingsphere-proxy-frontend-mysql/src/main/java/org/apache/shardingsphere/proxy/frontend/mysql/command/query/binary/prepare/MySQLComStmtPrepareExecutor.java]: Can it be moved to ProjectionsContext static api?

Comment (tristaZero) [reply shardingsphere-proxy/shardingsphere-proxy-frontend/shardingsphere-proxy-frontend-mysql/src/main/java/org/apache/shardingsphere/proxy/frontend/mysql/command/query/binary/prepare/MySQLComStmtPrepareExecutor.java]: Hi, sorry for my late response. I mean could we do some improvement on this bug-fix?
Please look at `getColumnCount()` of `ShardingSphereResultSetMetaData`. Actually it does a similar thing with you. So can we leverage the `ProjectionsContext` to deal with this issue? 
Maybe first we can call `SQLStatementContextFactory` to get `SelectStatementContext` and `ProjectionsContext`.
Then, could you figure out whether `ProjectionsContext` can return the `actual projection count`?

BTW, the failed CI still needs your attention. Thanks, hope this PR merged ASAP.

Comment (donjuan6) [reply shardingsphere-proxy/shardingsphere-proxy-frontend/shardingsphere-proxy-frontend-mysql/src/main/java/org/apache/shardingsphere/proxy/frontend/mysql/command/query/binary/prepare/MySQLComStmtPrepareExecutor.java]: ok, thanks!

File: shardingsphere-infra/shardingsphere-infra-binder/src/main/java/org/apache/shardingsphere/infra/binder/segment/select/projection/ProjectionsContext.java [hunk 3]
```diff
@@ -38,18 +38,18 @@
 @Getter
 @ToString
 public final class ProjectionsContext {
-    
+
```
Comment (terrymanu) [inline shardingsphere-infra/shardingsphere-infra-binder/src/main/java/org/apache/shardingsphere/infra/binder/segment/select/projection/ProjectionsContext.java:41]: There is no change on the class ProjectionsContext.java beside the indent, could you revert the change of this class?

**Review by terrymanu — APPROVED** [review]

**Review by tristaZero — APPROVED** [review]
@donjuan6 Hey great.

**Pull request merged**